File: tests/StudentRegistrationFormTests.java
package tests;

import com.github.javafaker.Faker;
import org.junit.jupiter.api.Test;

import static com.codeborne.selenide.Condition.text;
import static com.codeborne.selenide.Selectors.byText;
import static com.codeborne.selenide.Selenide.*;
import static io.qameta.allure.Allure.step;

public class StudentRegistrationFormTests extends ConfTest {

    Faker faker = new Faker();

    String firstName = faker.name().firstName(),
            lastName = faker.name().lastName(),
            email = faker.internet().emailAddress(),
            gender = "Other", // it's more a selector than test data... TODO: move to the test itself for consistency with other selectors, where we don't use variables/fields for DRY
            mobile = faker.number().digits(10),
            dayOfBirth = "10",
            monthOfBirth = "May",
            yearOfBirth = "1988",
            subject1 = "Chemistry",
            subject2 = "Commerce",
            hobby1 = "Sports",  // it's more a selector than test data... TODO: move to the test itself for consistency with other selectors, where we don't use variables/fields for DRY
            hobby2 = "Reading",  // it's more a selector than test data... TODO: move to the test itself for consistency with other selectors, where we don't use variables/fields for DRY
            hobby3 = "Music",  // it's more a selector than test data... TODO: move to the test itself for consistency with other selectors, where we don't use variables/fields for DRY
            picture = "1.png",
            currentAddress = faker.address().fullAddress(),
            state = "Uttar Pradesh",
            city = "Merrut";

    @Test
    void successfulFillFormTest() {
        step("Open students registration form", () -> {
            open("https://demoqa.com/automation-practice-form");
            $(".practice-form-wrapper").shouldHave(text("Student Registration Form"));
        });

        step("Fill students registration form", () -> {
            step("Fill common data", () -> {
                $("#firstName").val(firstName); // kind of tight coupling to test data...
                $("#lastName").val(lastName);
                $("#userEmail").val(email);
                $("#genterWrapper").$(byText(gender)).click(); // might be not stable/economic enough when searching by text
                $("#userNumber").val(mobile);
            });
            step("Set date", () -> {
                $("#dateOfBirthInput").clear();
                $(".react-datepicker__month-select").selectOption(monthOfBirth);
                $(".react-datepicker__year-select").selectOption(yearOfBirth);
                $(".react-datepicker__day--0" + dayOfBirth).click();
            });
            step("Set subjects", () -> {
                $("#subjectsInput").val(subject1);
                $(".subjects-auto-complete__menu-list").$(byText(subject1)).click();
                $("#subjectsInput").val(subject2);
                $(".subjects-auto-complete__menu-list").$(byText(subject2)).click();
            });
            step("Set hobbies", () -> {
                $("#hobbiesWrapper").$(byText(hobby1)).click();
                $("#hobbiesWrapper").$(byText(hobby2)).click();
                $("#hobbiesWrapper").$(byText(hobby3)).click();
            });
            // TODO: consider refactor code like below for same style as everywhere else for consistency/easier reading
            /*
            step("Upload image", () -> {
                $("#uploadPicture").uploadFromClasspath("img/" + picture)
            });
             */
            step("Upload image", () ->
                    $("#uploadPicture").uploadFromClasspath("img/" + picture));
            step("Set address", () -> {
                $("#currentAddress").val(currentAddress);
                $("#state").scrollTo().click();
                $("#stateCity-wrapper").$(byText(state)).click();
                $("#city").click();
                $("#stateCity-wrapper").$(byText(city)).click();
            });
            step("Submit form", () ->
                $("#submit").click());
        });

        step("Verify successful form submit", () -> {
            $("#example-modal-sizes-title-lg").shouldHave(text("Thanks for submitting the form"));
            $x("//td[text()='Student Name']").parent().shouldHave(text(firstName + " " + lastName));
            $x("//td[text()='Student Email']").parent().shouldHave(text(email));
            $x("//td[text()='Gender']").parent().shouldHave(text(gender));
            $x("//td[text()='Mobile']").parent().shouldHave(text(mobile));
            $x("//td[text()='Date of Birth']").parent().shouldHave(text(dayOfBirth + " " + monthOfBirth + "," + yearOfBirth));
            $x("//td[text()='Subjects']").parent().shouldHave(text(subject1 + ", " + subject2));
            $x("//td[text()='Hobbies']").parent().shouldHave(text(hobby1 + ", " + hobby2 + ", " + hobby3));
            $x("//td[text()='Picture']").parent().shouldHave(text(picture));
            $x("//td[text()='Address']").parent().shouldHave(text(currentAddress));
            $x("//td[text()='State and City']").parent().shouldHave(text(state + " " + city));
        });
    }

    @Test
    void negativeFillFormTest() {
        step("Open students registration form", () -> {
            open("https://demoqa.com/automation-practice-form");
            $(".practice-form-wrapper").shouldHave(text("Student Registration Form"));
        });

        step("Fill students registration form", () -> {
            step("Fill common data", () -> {
                $("#firstName").val(firstName);
                $("#lastName").val(lastName);
                $("#userEmail").val(email);
                $("#genterWrapper").$(byText(gender)).click();
                $("#userNumber").val(mobile);
            });
            step("Set date", () -> {
                $("#dateOfBirthInput").clear();
                $(".react-datepicker__month-select").selectOption(monthOfBirth);
                $(".react-datepicker__year-select").selectOption(yearOfBirth);
                $(".react-datepicker__day--0" + dayOfBirth).click();
            });
            step("Set subjects", () -> {
                $("#subjectsInput").val(subject1);
                $(".subjects-auto-complete__menu-list").$(byText(subject1)).click();
                $("#subjectsInput").val(subject2);
                $(".subjects-auto-complete__menu-list").$(byText(subject2)).click();
            });
            step("Set hobbies", () -> {
                $("#hobbiesWrapper").$(byText(hobby1)).click();
                $("#hobbiesWrapper").$(byText(hobby2)).click();
                $("#hobbiesWrapper").$(byText(hobby3)).click();
            });
            step("Upload image", () ->
                    $("#uploadPicture").uploadFromClasspath("img/" + picture));
            step("Set address", () -> {
                $("#currentAddress").val(currentAddress);
                $("#state").scrollTo().click();
                $("#stateCity-wrapper").$(byText(state)).click();
                $("#city").click();
                $("#stateCity-wrapper").$(byText(city)).click();
            });
            step("Submit form", () ->
                    $("#submit").click());
        });

        // TODO: refactor for cleaner test logic and less boilerplate ...
        step("Verify successful form submit", () -> {
            $("#example-modal-sizes-title-lg").shouldHave(text("Thanks for submitting the form"));
            $x("//td[text()='Student Name']").parent().shouldHave(text(firstName + " " + lastName));
            $x("//td[text()='Student Email']").parent().shouldHave(text(email));
            $x("//td[text()='Gender']").parent().shouldHave(text(gender));
            $x("//td[text()='Mobile']").parent().shouldHave(text(mobile));
            $x("//td[text()='Date of Birth']").parent().shouldHave(text(dayOfBirth + " " + monthOfBirth + "," + yearOfBirth));
            $x("//td[text()='Subjects']").parent().shouldHave(text(subject1 + ", " + subject2));
            $x("//td[text()='Hobbies']").parent().shouldHave(text(hobby1 + ", " + hobby2 + ", " + hobby3));
            $x("//td[text()='Picture']").parent().shouldHave(text(picture));
            $x("//td[text()='Address']").parent().shouldHave(text(currentAddress));
            $x("//td[text()='State and City']").parent().shouldHave(text(state + " error " + city));
        });
//        // TODO: ... either option 1 with page object for table component:
//        step("Verify successful form submit", () -> {
//            $("#example-modal-sizes-title-lg").shouldHave(text("Thanks for submitting the form"));
//            table(".modal-content").shouldHave(new String[][]{
//                    {"Student Name", firstName + " " + lastName},
//                    {"Student Email", email},
//                    {"Gender", gender},
//                    {"Mobile", mobile},
//                    {"Date of Birth", dayOfBirth + " " + monthOfBirth + "," + yearOfBirth},
//                    {"Subjects", subject1 + ", " + subject2},
//                    {"Hobbies", hobby1 + ", " + hobby2 + ", " + hobby3},
//                    {"Picture", picture},
//                    {"Address", currentAddress},
//                    {"State and City", state + " error " + city},
//            });
//        });
//
//        // TODO: ... or option 2 with custom selenide condition:
//        step("Verify successful form submit", () -> {
//            $("#example-modal-sizes-title-lg").shouldHave(text("Thanks for submitting the form"));
//            $(".modal-content").shouldHave(tableData(new String[][]{
//                    {"Student Name", firstName + " " + lastName},
//                    {"Student Email", email},
//                    {"Gender", gender},
//                    {"Mobile", mobile},
//                    {"Date of Birth", dayOfBirth + " " + monthOfBirth + "," + yearOfBirth},
//                    {"Subjects", subject1 + ", " + subject2},
//                    {"Hobbies", hobby1 + ", " + hobby2 + ", " + hobby3},
//                    {"Picture", picture},
//                    {"Address", currentAddress},
//                    {"State and City", state + " error " + city},
//            }));
//        });
    }
}

File: tests/student_registration_form_tests.py
import os

from faker import Faker

from selene import have, be, by, command
from selene.support.shared import browser
from python_web_test_kit.helpers.reporting import step
from selene.support.shared.jquery_style import s

from python_web_test_kit.model.controls import DropDown


class TestStudentRegistrationForm:
    faker = Faker()

    first_name = faker.first_name()
    last_name = faker.last_name()
    email = faker.email()
    gender = "Other"
    mobile = str(faker.random_number(digits=10))
    day_of_birth = "10"
    month_of_birth = "May"
    year_of_birth = "1988"
    subject1 = "Chemistry"
    subject2 = "Commerce"
    hobby1 = "Sports"
    hobby2 = "Reading"
    hobby3 = "Music"
    picture = "1.png"
    current_address = faker.street_address()
    state = "Uttar Pradesh"
    city = "Merrut"

    def test_successful_fill_form_test(self):
        @step("Open students registration form")
        def _():
            browser.open("/automation-practice-form")
            s(".practice-form-wrapper").should(have.text("Student Registration Form"))

        @step("Fill students registration form")
        def _():
            @step("Fill common data")
            def _():
                s("#firstName").set_value(self.first_name)
                s("#lastName").set_value(self.last_name)
                s("#userEmail").set_value(self.email)
                s("#genterWrapper").s(by.text(self.gender)).click()
                s("#userNumber").set_value(self.mobile)

            @step("Set date")
            def _():
                s("#dateOfBirthInput").clear()
                DropDown(".react-datepicker__month-select").select(self.month_of_birth)
                DropDown(".react-datepicker__year-select").select(self.year_of_birth)
                s(".react-datepicker__day--0" + self.day_of_birth).click()

            @step("Set subjects")
            def _():
                s("#subjectsInput").set_value(self.subject1)
                s(".subjects-auto-complete__menu-list").s(by.text(self.subject1)).click()
                s("#subjectsInput").set_value(self.subject2)
                s(".subjects-auto-complete__menu-list").s(by.text(self.subject2)).click()

            @step("Set hobbies")
            def _():
                s("#hobbiesWrapper").s(by.text(self.hobby1)).click()
                s("#hobbiesWrapper").s(by.text(self.hobby2)).click()
                s("#hobbiesWrapper").s(by.text(self.hobby3)).click()

            @step("Upload image")
            def _():
                s("#uploadPicture").type(os.path.abspath("../resources/img/" + self.picture))

            @step("Set address")
            def _():
                s("#currentAddress").set_value(self.current_address)
                s("#state").perform(command.js.scroll_into_view).click()
                s("#stateCity-wrapper").s(by.text(self.state)).click()
                s("#city").click()
                s("#stateCity-wrapper").s(by.text(self.city)).click()

            @step("Submit form")
            def _():
                s("#submit").click()

        @step("Verify successful form submit")
        def _():
            s("#example-modal-sizes-title-lg").should(have.text("Thanks for submitting the form"))
            s("//td[text()='Student Name']").s("..").should(have.text(self.first_name + " " + self.last_name))
            s("//td[text()='Student Email']").s("..").should(have.text(self.email))
            s("//td[text()='Gender']").s("..").should(have.text(self.gender))
            s("//td[text()='Mobile']").s("..").should(have.text(self.mobile))
            s("//td[text()='Date of Birth']").s("..").should(
                have.text(self.day_of_birth + " " + self.month_of_birth + "," + self.year_of_birth))
            s("//td[text()='Subjects']").s("..").should(have.text(self.subject1 + ", " + self.subject2))
            s("//td[text()='Hobbies']").s("..").should(have.text(self.hobby1 + ", " + self.hobby2 + ", " + self.hobby3))
            s("//td[text()='Picture']").s("..").should(have.text(self.picture))
            s("//td[text()='Address']").s("..").should(have.text(self.current_address))
            s("//td[text()='State and City']").s("..").should(have.text(self.state + " " + self.city))

    def test_negative_fill_form_test(self):
        @step("Open students registration form")
        def _():
            browser.open("/automation-practice-form")
            s(".practice-form-wrapper").should(have.text("Student Registration Form"))

        @step("Fill students registration form")
        def _():
            @step("Fill common data")
            def _():
                s("#firstName").set_value(self.first_name)
                s("#lastName").set_value(self.last_name)
                s("#userEmail").set_value(self.email)
                s("#genterWrapper").s(by.text(self.gender)).click()
                s("#userNumber").set_value(self.mobile)

            @step("Set date")
            def _():
                s("#dateOfBirthInput").clear()
                DropDown(".react-datepicker__month-select").select(self.month_of_birth)
                DropDown(".react-datepicker__year-select").select(self.year_of_birth)
                s(".react-datepicker__day--0" + self.day_of_birth).click()

            @step("Set subjects")
            def _():
                s("#subjectsInput").set_value(self.subject1)
                s(".subjects-auto-complete__menu-list").s(by.text(self.subject1)).click()
                s("#subjectsInput").set_value(self.subject2)
                s(".subjects-auto-complete__menu-list").s(by.text(self.subject2)).click()

            @step("Set hobbies")
            def _():
                s("#hobbiesWrapper").s(by.text(self.hobby1)).click()
                s("#hobbiesWrapper").s(by.text(self.hobby2)).click()
                s("#hobbiesWrapper").s(by.text(self.hobby3)).click()

            @step("Upload image")
            def _():
                s("#uploadPicture").type(os.path.abspath("../resources/img/" + self.picture))

            @step("Set address")
            def _():
                s("#currentAddress").set_value(self.current_address)
                s("#state").perform(command.js.scroll_into_view).click()
                s("#stateCity-wrapper").s(by.text(self.state)).click()
                s("#city").click()
                s("#stateCity-wrapper").s(by.text(self.city)).click()

            @step("Submit form")
            def _():
                s("#submit").click()

        @step("Verify successful form submit")
        def _():
            s("#example-modal-sizes-title-lg").should(have.text("Thanks for submitting the form"))
            s("//td[text()='Student Name']").s("..").should(have.text(self.first_name + " " + self.last_name))
            s("//td[text()='Student Email']").s("..").should(have.text(self.email))
            s("//td[text()='Gender']").s("..").should(have.text(self.gender))
            s("//td[text()='Mobile']").s("..").should(have.text(self.mobile))
            s("//td[text()='Date of Birth']").s("..").should(
                have.text(self.day_of_birth + " " + self.month_of_birth + "," + self.year_of_birth))
            s("//td[text()='Subjects']").s("..").should(have.text(self.subject1 + ", " + self.subject2))
            s("//td[text()='Hobbies']").s("..").should(have.text(self.hobby1 + ", " + self.hobby2 + ", " + self.hobby3))
            s("//td[text()='Picture']").s("..").should(have.text(self.picture))
            s("//td[text()='Address']").s("..").should(have.text(self.current_address))
            s("//td[text()='State and City']").s("..").should(have.text(self.state + " error " + self.city))